File: Assets/EchoSim/Simulation/Agents/AgentMind.cs
using System;
using System.Collections.Generic;
using EchoSim.Core;

namespace EchoSim.Simulation
{
    /// <summary>
    /// Per-resident cognition state: personality, needs, current goal commitment.
    /// Owned by the resident registry; never serialized as-is (Sprint 18 adds DTOs).
    /// </summary>
    public sealed class AgentMind
    {
        private readonly Dictionary<GoalId, SimTime> _lastSelectedAt = new Dictionary<GoalId, SimTime>();

        public AgentId Agent { get; }
        public PersonalityProfile Personality { get; }
        public NeedSet Needs { get; }

        /// <summary>Goal currently committed (may differ from raw winner due to hysteresis).</summary>
        public GoalDefinition? CommittedGoal { get; internal set; }

        public bool HasCommitment => CommittedGoal != null;

        public GoalId? CurrentGoalId => CommittedGoal?.Id;

        /// <summary>Assigned job (Sprint 5); null for unemployed residents.</summary>
        public JobDefinition? Job { get; internal set; }

        /// <summary>Personal weekly rhythm (Sprint 5); null when fully utility-driven.</summary>
        public WeeklySchedule? Schedule { get; internal set; }

        /// <summary>Seeded routine jitter in minutes (spec §5.6).</summary>
        public int RoutineOffsetMinutes { get; private set; }

        /// <summary>Authored likes/dislikes (Sprint 14); 0 when unspecified.</summary>
        public PreferenceProfile Preferences { get; private set; } = new PreferenceProfile();

        /// <summary>Wallet (Sprint 15). Mutate via systems, not directly.</summary>
        public float Money { get; set; }

        public Inventory Inventory { get; } = new Inventory();

        public void SetSchedule(WeeklySchedule? schedule) => Schedule = schedule;
        public void SetRoutineOffset(int offsetMinutes) => RoutineOffsetMinutes = offsetMinutes;
        public void SetPreferences(PreferenceProfile preferences) =>
            Preferences = preferences ?? throw new ArgumentNullException(nameof(preferences));

        internal AgentMind(AgentId agent, PersonalityProfile personality, NeedSet needs)
        {
            Agent = agent;
            Personality = personality ?? throw new ArgumentNullException(nameof(personality));
            Needs = needs ?? throw new ArgumentNullException(nameof(needs));
        }

        public IReadOnlyDictionary<GoalId, SimTime> LastSelectedAt => _lastSelectedAt;

        internal void NoteSelection(GoalId goal, SimTime at) => _lastSelectedAt[goal] = at;

        internal void ReleaseCommitment() => CommittedGoal = null;

        /// <summary>
        /// Commitment hysteresis: a committed satisfier goal is kept while any of its
        /// relief needs remain above the satisfaction threshold (prevents flapping).
        /// </summary>
        public bool CommitmentHolds()
        {
            if (CommittedGoal == null) return false;
            var reliefs = CommittedGoal.ReliefNeeds;
            for (int i = 0; i < reliefs.Count; i++)
            {
                if (!Needs.Get(reliefs[i]).IsSatisfied) return true;
            }
            // All relief needs satisfied: release and remember for cooldown purposes.
            CommittedGoal = null;
            return false;
        }

        public float EmotionValence { get; private set; }

        /// <summary>Sets mood valence (-1..1); clamped (Sprint 8 writes via EmotionSystem).</summary>
        public void SetEmotion(float valence) => EmotionValence = Math.Clamp(valence, -1f, 1f);

        /// <summary>
        /// Persistent planner facts for this resident (has_meal, has_ingredients, ...).
        /// Written back by plan execution so future replans observe past results.
        /// Ephemeral completion flags are reset per replan by the state builder.
        /// </summary>
        public IDictionary<string, int> PlannerMemory { get; } = new Dictionary<string, int>(StringComparer.Ordinal);
    }

    /// <summary>Authored description used to spawn a full resident.</summary>
    public sealed class ResidentSpec
    {
        public string Id { get; }
        public string DisplayName { get; }
        public string? HomeLocationId { get; set; }
        public string? StartLocationId { get; set; }
        public PersonalityProfile Personality { get; set; } = PersonalityProfile.Balanced();
        public IDictionary<NeedKind, float>? InitialNeeds { get; set; }
        public bool UseStandardNeeds { get; set; } = true;

        public ResidentSpec(string id, string displayName)
        {
            Id = id ?? throw new ArgumentNullException(nameof(id));
            DisplayName = displayName ?? throw new ArgumentNullException(nameof(displayName));
        }
    }

    /// <summary>Ordered registry of resident minds.</summary>
    public sealed class ResidentRegistry
    {
        private readonly Dictionary<AgentId, AgentMind> _minds = new Dictionary<AgentId, AgentMind>();
        private readonly List<AgentId> _order = new List<AgentId>();

        public IReadOnlyList<AgentId> OrderedIds => _order;
        public int Count => _order.Count;

        public void Add(AgentMind mind)
        {
            _minds.Add(mind.Agent, mind);
            _order.Add(mind.Agent);
        }

        public bool TryGet(AgentId id, out AgentMind mind) => _minds.TryGetValue(id, out mind!);

        public AgentMind Get(AgentId id)
        {
            if (!_minds.TryGetValue(id, out var mind))
                throw new KeyNotFoundException($"Unknown resident '{id}'.");
            return mind!;
        }

        public IEnumerable<AgentMind> AllInOrder()
        {
            for (int i = 0; i < _order.Count; i++) yield return _minds[_order[i]];
        }
    }
}

File: packages/cognition/src/planning/agentMind.ts
/** Per-resident cognition state: personality, needs, goal commitment, mood. */
import type { NeedSet } from "../needs/needs.js";
import type { PersonalityProfile } from "../personality/personality.js";
import type { GoalDefinition, PreferenceProfileLike } from "../goals/goalModel.js";
import { EPHEMERAL_FACTS } from "./facts.js";

export class AgentMind {
  committedGoal: GoalDefinition | null = null;
  emotionValence = 0;
  money = 0;
  routineOffsetMinutes = 0;
  readonly inventory = new Map<string, number>();
  readonly plannerMemory = new Map<string, number>();
  private readonly lastSelected = new Map<string, number>();
  preferences: PreferenceProfileLike = { get: () => 0 };

  constructor(
    readonly agent: string,
    readonly displayName: string,
    readonly homeLocationId: string | undefined,
    readonly personality: PersonalityProfile,
    readonly needs: NeedSet,
  ) {}

  get hasCommitment(): boolean { return this.committedGoal !== null; }
  get currentGoalId(): string | undefined { return this.committedGoal?.id ?? undefined; }

  noteSelection(goal: string, atMinutes: number): void {
    this.lastSelected.set(goal, atMinutes);
  }
  lastSelectedSnapshot(): ReadonlyMap<string, number> {
    return this.lastSelected;
  }
  releaseCommitment(): void {
    this.committedGoal = null;
  }
  setEmotion(valence: number): void {
    if (Number.isNaN(valence)) throw new Error("valence must be a number");
    this.emotionValence = Math.max(-1, Math.min(1, valence));
  }
  setPreferences(preferences: PreferenceProfileLike | undefined): void {
    this.preferences = preferences ?? { get: () => 0 };
  }

  /**
   * Commitment hysteresis: kept while any relief need is still above its
   * satisfaction threshold; releases (returns false) once fully satisfied.
   */
  commitmentHolds(): boolean {
    if (!this.committedGoal) return false;
    for (const kind of this.committedGoal.reliefNeeds) {
      if (!this.needs.get(kind).isSatisfied) return true;
    }
    this.committedGoal = null;
    return false;
  }
}

export interface ResidentSpec {
  id: string;
  displayName: string;
  homeLocationId?: string;
  startLocationId?: string;
  personality: import("../personality/personality.js").PersonalityProfile;
  initialNeeds?: Partial<Record<number, number>>;
  preferences?: PreferenceProfileLike;
}

/** Persistent non-location, non-ephemeral fact write-back used by execution. */
export function applyEffectToMemory(
  memory: Map<string, number>,
  effect: { mode: "assign"; key: string; value: number } | { mode: "add"; key: string; delta: number },
): void {
  if (effect.key.startsWith("at_")) return;
  if (EPHEMERAL_FACTS.has(effect.key)) return;
  const current = memory.get(effect.key) ?? 0;
  memory.set(effect.key, effect.mode === "assign" ? effect.value : current + effect.delta);
}

/** Copies persistent facts into a fresh planner view; ephemerals start at 0. */
export function copyPlannerMemoryInto(state: { set(key: string, value: number): void }, memory: ReadonlyMap<string, number>): void {
  for (const [k, v] of memory) {
    if (k.startsWith("at_")) continue;
    if (EPHEMERAL_FACTS.has(k)) continue;
    state.set(k, v);
  }
  for (const key of EPHEMERAL_FACTS) state.set(key, 0);
}
